modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/portlet/action/CopyFileEntryMVCActionCommand.java

```diff
@@ -116,9 +138,166 @@ private void _copyFileEntry(
 		}
 	}
 
+	private ServiceContext _createServiceContext(
```

Comment:
It is possible that `_createServiceContext`  returns a partially initialized service context when an exception is thrown. To avoid that, can you make the happy path and the error path independent?
```
private ServiceContext _createServiceContext(
			ActionRequest actionRequest, long fileEntryId, Group group)
		throws PortalException {

		try {
			AssetEntry assetEntry = _assetEntryService.getEntry(
				DLFileEntry.class.getName(), fileEntryId);

			List<Long> groupIds = _getGroupIds(
				ListUtil.fromArray(
					_portal.getCurrentAndAncestorSiteGroupIds(
						group.getGroupId())),
				group);

			ServiceContext serviceContext = ServiceContextFactory.getInstance(
				DLFileEntry.class.getName(), actionRequest);
			
			serviceContext.setAssetCategoryIds(
				_getAssetCategoryIds(assetEntry, group, groupIds));

			serviceContext.setAssetTagNames(
				_getAssetTagNames(assetEntry, groupIds));

			return serviceContext;
		}
		catch (NoSuchEntryException noSuchEntryException) {
			return ServiceContextFactory.getInstance(
				DLFileEntry.class.getName(), actionRequest);
		}
	}
```